# ID2214 Assignment 1 Group no. [7]
### Project members: 
[Valgerdur Tryggvadottir, vtry@kth.se]
[Cecilia Battinelli, cbat@kth.se]
[Parastu Rahgozar, parastu@kth.se]
[Zhiwu Dong, zhiwu@kth.se]

### Declaration:
By submitting this solution, it is hereby declared that all individuals listed above have contributed to the solution, either with code that appear in the final solution below, or with code that has been evaluated and compared to the final solution, but for some reason has been excluded. It is also declared that all project members fully understand all parts of the final solution and can explain it upon request.

It is furthermore declared that the code below is a contribution by the project members only, and specifically that no part of the solution has been copied from any other source (except for lecture slides at the course ID2214) and no part of the solution has been provided by someone not listed as project member above.

It is furthermore declared that it has been understood that no other library/package than the Python 3 standard library, NumPy and pandas may be used in the solution for this assignment.

### Instructions
All assignments starting with number 1 below are mandatory. Satisfactory solutions
will give 1 point (in total). If they in addition are good (all parts work more or less 
as they should), completed on time (submitted before the deadline in Canvas) and according
to the instructions, together with satisfactory solutions of assignments starting with 
number 2 below, then the assignment will receive 2 points (in total).

It is highly recommended that you do not develop the code directly within the notebook
but that you copy the comments and test cases to your regular development environment
and only when everything works as expected, that you paste your functions into this
notebook, do a final testing (all cells should succeed) and submit the whole notebook 
(a single file) in Canvas (do not forget to fill in your group number and names above).


## Load NumPy and pandas

In [40]:
import numpy as np
import pandas as pd

## 1a. Create and apply normalization

In [41]:
# Insert the functions create_normalization and apply_normalization below (after the comments)
#
# Input to create_normalization:
# df: a dataframe (where the column names "CLASS" and "ID" have special meaning)
# normalizationtype: "minmax" (default) or "zscore"
#
# Output from create_normalization:
# df: a new dataframe, where each numeric value in a column has been replaced by a normalized value
# normalization: a mapping (dictionary) from each column name to a triple, consisting of
#                ("minmax",min_value,max_value) or ("zscore",mean,std)
#
# Hint 1: First copy the input dataframe and modify the copy (the input dataframe should be kept unchanged)
# Hint 2: Consider columns of type "float" or "int" only (and which are not labeled "CLASS" or "ID"),
#         the other columns should remain unchanged
# Hint 3: Take a close look at the lecture slides on data preparation
#
# Input to apply_normalization:
# df: a dataframe
# normalization: a mapping (dictionary) from column names to triples (see above)
#
# Output from apply_normalization:
# df: a new dataframe, where each numerical value has been normalized according to the mapping
#
# Hint 1: First copy the input dataframe and modify the copy (the input dataframe should be kept unchanged)
# Hint 2: For minmax-normalization, you may consider to limit the output range to [0,1]

def create_normalization(df,normalizationtype="minmax"):
    d = {}
    df1 = df.copy()
    # Go through the columns different from ID and CLASS
    # pick the max and min value and append them (tog. with the norm_type name)
    # to a dictionary d whose key is the column name.
    # Then go through all the elements in the column in df1 and normalize them
    # finally change the column values using the normalized ones
    if normalizationtype=="minmax":
        for elem in df1.columns:
            if elem!='ID' and elem!='CLASS':
                minn = df1[elem].min()
                maxx = df1[elem].max()
                d[elem] = ("minmax", minn, maxx)
                df1[elem] = [(x-minn)/(maxx-minn) for x in df[elem]]
    else:
    # Go through the columns different from ID and CLASS
    # find the mean and std value and append them (tog. with the norm_type name)
    # to a dictionary whose key is the column name.
    # Then go through all the elements in the column and normalize them-
    # Finally change the column values using the normalized ones
        for elem in df1.columns:
            if elem!='ID' and elem!='CLASS':
                mean = df1[elem].mean()
                std = df1[elem].std()
                d[elem] = ("zscore",mean,std)
                df1[elem] = df1[elem].apply(lambda x: (x-mean)/std)
    return df1, d


def apply_normalization(df,normalization):
    df1 = df.copy()
    # Go through the columns different from ID and CLASS
    # and normalize the column's value, according to the chosen method,
    # using the referring values from mapping dictionary normalization
    # (i.e. min & max for minmax & mean, std for zscore)
    # NOTE: for minmax normalization type normalized values > 1 are set to 1
    # & normalized values < 0 are set to 0 (Normalization minmax must range 0-1)
    # Finally change the column's values using the normalized ones
    for elem in df1.columns:
        if elem!='ID' and elem!='CLASS':
            values = df1[elem].values
            tup_elem = normalization[elem]
            for i in range(len(values)):
                if tup_elem[0] == "minmax":
                    values[i] =  (values[i] - tup_elem[1])/(tup_elem[2]-tup_elem[1])
                    #Here we consider only into 0-1 interval
                    if values[i] < 0.0:
                        values[i] = 0.0
                    if values[i] > 1.0:
                        values[i] = 1.0
                    df1[elem] = values
                elif tup_elem[0] == "zscore":
                    values[i] = (values[i] - tup_elem[1]) / tup_elem[2]
                    df1[elem] = values

    return df1


In [42]:
# Test your code (leave this part unchanged)

glass_train_df = pd.read_csv("glass_train.txt")

glass_test_df = pd.read_csv("glass_test.txt")

glass_train_norm, normalization = create_normalization(glass_train_df,normalizationtype="minmax")
print("normalization:\n")
for f in normalization:
    print("{}:{}".format(f,normalization[f]))

glass_test_norm = apply_normalization(glass_test_df,normalization)
print("\nglass_test_norm:\n")
glass_test_norm

normalization:

RI:('minmax', 1.51131, 1.53125)
Na:('minmax', 10.73, 15.79)
Mg:('minmax', 0.0, 4.49)
Al:('minmax', 0.29, 3.04)
Si:('minmax', 69.81, 75.18)
K:('minmax', 0.0, 6.21)
Ca:('minmax', 5.43, 14.68)
Ba:('minmax', 0.0, 3.15)
Fe:('minmax', 0.0, 0.37)

glass_test_norm:



ID        RI        Na        Mg        Al        Si         K  \
0    101  0.262788  0.399209  0.634744  0.418182  0.644320  0.091787   
1    104  0.799398  0.606719  0.701559  0.134545  0.141527  0.012882   
2     44  0.541123  0.592885  0.855234  0.156364  0.363128  0.027375   
3     17  0.327482  0.385375  0.817372  0.316364  0.614525  0.098229   
4     81  0.231194  0.420949  0.783964  0.665455  0.530726  0.111111   
5    142  0.361083  0.488142  0.808463  0.283636  0.562384  0.091787   
6    120  0.261284  0.559289  0.795100  0.429091  0.491620  0.103060   
7    123  0.278837  0.494071  0.788419  0.432727  0.564246  0.090177   
8    133  0.342026  0.533597  0.886414  0.323636  0.499069  0.093398   
9    185  0.000000  1.000000  0.000000  0.018182  1.000000  0.000000   
10   212  0.468405  0.717391  0.000000  0.629091  0.672253  0.000000   
11   199  0.200602  0.721344  0.000000  0.861818  0.612663  0.006441   
12    83  0.258275  0.529644  0.790646  0.349091  0.558659  0.109501   
13   158  0.496489  0.652174  0.837416  0.105455  0.368715  0.017713   
14    42  0.312939  0.391304  0.761693  0.330909  0.631285  0.095008   
15   153  0.324975  0.575099  0.812918  0.130909  0.594041  0.009662   
16    40  0.542628  0.687747  0.850780  0.065455  0.364991  0.017713   
17    90  0.255266  0.359684  0.775056  0.574545  0.636872  0.101449   
18   149  0.270311  0.496047  0.795100  0.396364  0.538175  0.090177   
19   194  0.294885  0.794466  0.000000  0.621818  0.597765  0.000000   
20   164  0.192076  0.648221  0.596882  1.000000  0.014898  0.270531   
21   134  0.335507  0.588933  0.875278  0.454545  0.372439  0.086957   
22   128  0.476429  0.602767  0.507795  0.414545  0.405959  0.078905   
23   137  0.338516  0.448617  0.846325  0.287273  0.607076  0.090177   
24   136  0.329990  0.486166  0.868597  0.367273  0.469274  0.088567   
25     8  0.313440  0.478261  0.804009  0.276364  0.638734  0.091787   
26    33  0.322969  0.418972  0.775056  0.341818  0.588454  0.098229   
27   162  0.402708  0.575099  0.788419  0.167273  0.528864  0.025765   
28   157  0.262788  0.529644  0.755011  0.360000  0.527002  0.083736   
29    72  0.359579  0.575099  0.861915  0.356364  0.400372  0.086957   
..   ...       ...       ...       ...       ...       ...       ...   
77    97  0.356068  0.452569  0.806236  0.280000  0.471136  0.103060   
78   129  0.469910  0.557312  0.465479  0.501818  0.441341  0.085346   
79   108  1.000000  0.310277  0.000000  0.258182  0.065177  0.019324   
80   168  0.420261  0.377470  0.000000  0.494545  0.733706  0.061192   
81    79  0.241725  0.630435  0.783964  0.349091  0.571695  0.059581   
82    63  0.522066  0.549407  0.859688  0.214545  0.368715  0.037037   
83    10  0.312939  0.448617  0.801782  0.389091  0.592179  0.091787   
84    35  0.326981  0.387352  0.788419  0.381818  0.584730  0.091787   
85    61  0.388164  0.567194  0.806236  0.298182  0.527002  0.022544   
86    29  0.319458  0.361660  0.783964  0.414545  0.621974  0.091787   
87   208  0.351053  0.723320  0.000000  0.556364  0.567970  0.227053   
88   100  0.341023  0.440711  0.659243  0.414545  0.579143  0.096618   
89   187  0.354564  0.709486  0.726058  0.701818  0.268156  0.235105   
90   135  0.341023  0.513834  0.857461  0.349091  0.553073  0.083736   
91   115  0.359077  0.468379  0.884187  0.327273  0.489758  0.096618   
92   171  0.620863  0.535573  0.000000  0.469091  0.448790  0.051530   
93    19  0.391174  0.626482  0.830735  0.323636  0.430168  0.009662   
94     9  0.394684  0.654150  0.797327  0.392727  0.422719  0.090177   
95    38  0.334002  0.397233  0.775056  0.385455  0.586592  0.103060   
96   154  0.240221  0.531621  0.757238  0.338182  0.536313  0.095008   
97    69  0.512036  0.472332  0.797327  0.221818  0.445065  0.037037   
98    23  0.303410  0.405138  0.806236  0.363636  0.554935  0.095008   
99   196  0.207623  0.673913  0.000000  0.869091  0.666667  0.012882   
100   66  0.485456  0.584980  0.799555  0.3018

### Comment on assumptions, things that do not work properly, etc.


## 1b. Create and apply imputation

In [43]:
# Insert the functions create_imputation and apply_imputation below (after the comments)
#
# Input to create_imputation:
# df: a dataframe (where the column names "CLASS" and "ID" have special meaning)
#
# Output from create_imputation:
# df: a new dataframe, where each missing numeric value in a column has been replaced by the mean of that column 
#     and each missing categoric value in a column has been replaced by the mode of that column
# imputation: a mapping (dictionary) from column name to value that has replaced missing values
#
# Hint 1: First copy the input dataframe and modify the copy (the input dataframe should be kept unchanged)
# Hint 2: Handle columns of type "float" or "int" only (and which are not labeled "CLASS" or "ID") in one way
#         and columns of type "object" and "category" in other ways
# Hint 3: Consider using the pandas functions mean() and mode() respectively, as well as fillna
# Hint 4: In the rare case of all values in a column being missing, replace numeric values with 0,
#         object values with "" and category values with the first category (cat.categories[0])  
#
# Input to apply_imputation:
# df: a dataframe
# imputation: a mapping (dictionary) from column name to value that should replace missing values
#
# Output from apply_imputation:
# df: a new dataframe, where each missing value has been replaced according to the mapping
#
# Hint 1: First copy the input dataframe and modify the copy (the input dataframe should be kept unchanged)
# Hint 2: Consider using fillna

def create_imputation(df1):
    df = df1.copy()
    d = {}
    # Go through all the columns whose name is different from ID & CLASS
    # First we analyze numerical columns and if all the values are missing
    # we fill them with 0, otherwise we fill the missing values with the mean of the column.
    # We append to the dictionary the mean value of the column referring to the column
    # name as key.

    # Later(else) we go through object and category columns. If all values are missing
    # we fill them with "" for object column, or first element for category type column
    # otherwise we fill with the first element of mode array.
    # We append to the dictionary the mode value [0] of the column referring to the column
    # name as key.
    
    for elem in df.columns:
        if elem!='ID' and elem!='CLASS':
            if df[elem].dtype == "float64" or df[elem].dtype == "int64":
                values = df[elem].values     
                if np.all(np.isnan(values)):
                    df[elem].fillna(0, inplace = True)
                else:
                      df[elem].fillna(df[elem].mean(), inplace = True)
                    

                value = df[elem].mean() 
                d[elem] = value
            else:
                values = df[elem].values
                if values.all() == np.nan:
                    if df[elem].dtype == "object":
                        df[elem].fillna("", inplace = True)
                    elif df[elem].dtype == "category":
                        df[elem].fillna(df[elem].cat.categories[0], inplace = True)
                else:
                    df[elem].fillna(df[elem].mode()[0], inplace = True)

                value2 = df[elem].mode()[0]
                d[elem] = value2

    return df, d


def apply_imputation(df1,imputation):
    df = df1.copy()
    
    # Go through all columns whose name is not ID or CLASS and fill the missing
    # values reffering to the imputation dictionary, i.e, given the column name, we fill
    # the missing value using the valueZ in dictionary (elem: valueZ).

    for elem in df1.columns:
        if elem!='ID' and elem!='CLASS':
            tup_elem = imputation[elem]
            df[elem].fillna(tup_elem, inplace = True)

    return df

In [44]:
# Test your code (leave this part unchanged)

anneal_train_df = pd.read_csv("anneal_train.txt")
anneal_test_df = pd.read_csv("anneal_test.txt")

anneal_train_imp, imputation = create_imputation(anneal_train_df)
anneal_test_imp = apply_imputation(anneal_test_df,imputation)

print("Imputation:\n")
for f in imputation:
    print("{}:{}".format(f,imputation[f]))

print("\nNo. of replaced missing values in training data:\n{}".format(anneal_train_imp.count()-anneal_train_df.count()))
print("\nNo. of replaced missing values in test data:\n{}".format(anneal_test_imp.count()-anneal_test_df.count()))



Imputation:

family:TN
product-type:C
steel:A
carbon:3.859688195991091
hardness:13.084632516703786
temper_rolling:T
condition:S
formability:2.2517482517482557
strength:26.302895322939868
non-ageing:N
surface-finish:P
surface-quality:E
enamelability:1.7142857142857018
bc:Y
bf:Y
bt:Y
bw/me:B
bl:Y
m:0.0
chrom:C
phos:P
cbond:Y
marvi:0.0
exptl:0.0
ferro:Y
corr:0.0
blue-bright-varn-clean:B
lustre:Y
jurofm:0.0
s:0.0
p:0.0
shape:SHEET
thick:1.1911937639198227
width:769.4917594654789
len:1229.293986636971
oil:Y
bore:35.18930957683742
packing:3.0

No. of replaced missing values in training data:
family                    382
product-type                0
steel                      43
carbon                      0
hardness                    0
temper_rolling            374
condition                 160
formability               163
strength                    0
non-ageing                391
surface-finish            444
surface-quality           128
enamelability             442
bc               

### Comment on assumptions, things that do not work properly, etc.

## 1c. Create and apply discretization

In [45]:
# Insert the functions create_bins and apply_bins below
#
# Input to create_bins:
# df: a dataframe
# nobins: no. of bins (default = 10)
# bintype: either "equal-width" (default) or "equal-size" 
#
# Output from create_bins:
# df: a new dataframe, where each numeric feature value has been replaced by a categoric (corresponding to some bin)
# binning: a mapping (dictionary) from column name to bins (threshold values for the bin)
#
# Hint 1: First copy the input dataframe and modify the copy (the input dataframe should be kept unchanged)
# Hint 2: Discretize columns of type "float" or "int" only (and which are not labeled "CLASS" or "ID")
# Hint 3: Consider using pd.cut and pd.qcut respectively, with labels=False and retbins=True
# Hint 4: Set all columns in the new dataframe to be of type "category"
# Hint 5: Set the categories of the discretized features to be [0,...,nobins-1]
# Hint 6: Change the first and the last element of each binning to -np.inf and np.inf respectively 
#
# Input to apply_bins:
# df: a dataframe
# binning: a mapping (dictionary) from column name to bins (threshold values for the bin)
#
# Output from apply_bins:
# df: a new dataframe, where each numeric feature value has been replaced by a categoric (corresponding to some bin)
#
# Hint 1: First copy the input dataframe and modify the copy (the input dataframe should be kept unchanged)
# Hint 2: Consider using pd.cut 
# Hint 3: Set all columns in the new dataframe to be of type "category"
# Hint 4: Set the categories of the discretized features to be [0,...,nobins-1]
#

def create_bins(dff,nobins=10,bintype="equal_width"):
    d = {}
    df = dff.copy()
    # Go through all the columns whose name is different from CLASS and ID
    # we create the numpy array of the column values and create the bins accordingly
    # to the selected method (i.e. bintype).
    # The bins are converted into a list and the first element is set to -inf and the last
    # to +inf. The bins are appendend to the dictionary referring to the column name as key.

    # The dataframe values are discretized according to the res  (df[elem])
    # and then the type is set to category type
    
    if bintype=="equal_width":
        for elem in df.columns:
            if elem!='ID' and elem!='CLASS':
                values = df[elem].values
                res, bins = pd.cut(values,nobins,retbins=True,labels=False)
                bins=list(bins)
                bins[0] = -np.inf
                bins[-1] = np.inf
                d[elem] = bins
                df[elem]= res
                df[elem] = df[elem].astype("category")
                df[elem].cat.categories
                lab = list(range(0,nobins))
                df[elem] = df[elem].cat.set_categories(lab)
    else:
        for elem in df.columns:
            if elem!='ID' and elem!='CLASS':
                values = df[elem].values
                res, bins = pd.qcut(values,nobins,retbins=True,labels=False, duplicates='drop')
                bins = list(bins)
                bins[0] = -np.inf
                bins[-1]= np.inf
                d[elem] =  bins
                df[elem] = res
                df[elem] = df[elem].astype("category")
                df[elem].cat.categories
                lab = list(range(0,nobins))
                df[elem] = df[elem].cat.set_categories(lab)


    return df, d

def apply_bins(dff,binning):
    df = dff.copy()
    # Go through all the columns whose name is different from CLASS and ID
    # and we set the values of the column categorized accordingly to the bins defined in the
    # mapping binning dictionary.
    # The categorized values are picked as the value of the column name-key in the dictionary (binning) that
    # is passed as argument.

    # Finally the type is set to category
    for elem in df.columns:
        if elem!='ID' and elem!='CLASS':
                values = df[elem].values
                bins = binning[elem]
                res = pd.cut(values,bins,labels=False)
                df[elem] = res
                df[elem] = df[elem].astype("category")
                df[elem].cat.categories
    return df


In [46]:
# Test your code  (leave this part unchanged)

glass_train_df = pd.read_csv("glass_train.txt")

glass_test_df = pd.read_csv("glass_test.txt")

glass_train_disc, binning = create_bins(glass_train_df,nobins=10,bintype="equal-size")
print("binning:\n")
for f in binning:
    print("{}:{}".format(f,binning[f]))

glass_test_disc = apply_bins(glass_test_df,binning)
print("\nglass_test_disc:\n")
glass_test_disc


binning:

RI:[-inf, 1.515896, 1.51618, 1.516516, 1.516866, 1.5175299999999998, 1.517902, 1.518618, 1.520114, 1.521846, inf]
Na:[-inf, 12.73, 12.872, 13.0, 13.222000000000001, 13.38, 13.492, 13.793999999999999, 14.197999999999999, 14.82, inf]
Mg:[-inf, 1.8200000000000003, 3.188, 3.41, 3.476, 3.55, 3.61, 3.728, inf]
Al:[-inf, 0.906, 1.172, 1.23, 1.3480000000000003, 1.48, 1.54, 1.622, 1.8080000000000003, 2.094, inf]
Si:[-inf, 71.756, 72.196, 72.388, 72.72, 72.79, 72.966, 73.06, 73.208, 73.372, inf]
K:[-inf, 0.006000000000000085, 0.14800000000000005, 0.39, 0.54, 0.5760000000000001, 0.6, 0.6360000000000002, 0.67, inf]
Ca:[-inf, 7.978, 8.112, 8.338, 8.554, 8.67, 8.81, 9.032, 9.674000000000001, 10.924000000000001, inf]
Ba:[-inf, 0.7800000000000004, inf]
Fe:[-inf, 0.06200000000000003, 0.1180000000000001, 0.24, inf]

glass_test_disc:



ID RI Na Mg Al Si  K Ca Ba Fe  CLASS
0    101  3  1  1  4  8  4  5  0  2      2
1    104  9  7  1  0  0  1  9  0  0      2
2     44  9  6  7  0  1  2  8  0  0      1
3     17  5  0  6  1  7  6  5  0  0      1
4     81  1  1  4  9  3  8  0  0  0      2
5    142  6  3  6  1  5  4  3  0  2      2
6    120  3  6  5  4  3  7  0  0  0      2
7    123  4  4  4  4  5  4  1  0  0      2
8    133  6  5  7  2  3  5  2  0  0      2
9    185  0  9  0  0  9  0  0  0  0      6
10   212  8  8  0  8  9  0  3  1  0      7
11   199  0  8  0  9  7  1  7  0  0      7
12    83  2  5  4  3  5  8  1  0  0      2
13   158  8  7  7  0  1  1  7  0  0      3
14    42  5  0  3  2  7  5  4  0  0      1
15   153  5  6  6  0  6  1  6  0  0      3
16    40  9  8  7  0  1  1  7  0  0      1
17    90  2  0  4  8  8  6  1  0  1      2
18   149  3  4  5  4  3  4  3  0  1      3
19   194  4  8  0  8  6  0  3  1  1      7
20   164  0  7  1  9  0  8  0  1  0      5
21   134  6  6  7  5  1  3  2  0  2      2
22   128  8  6  1  4  1  3  8  0  2      2
23   137  6  2  7  1  7  4  3  0  2      2
24   136  5  3  7  3  2  4  3  0  3      2
25     8  5  3  5  1  8  4  2  0  0      1
26    33  5  1  4  2  6  6  4  0  2      1
27   162  7  6  4  0  3  2  6  0  2      3
28   157  3  5  2  3  3  3  4  0  0      3
29    72  6  6  7  3  1  3  2  0  3      2
..   ... .. .. .. .. .. .. .. .. ..    ...
77    97  6  3  6  1  2  7  7  0  2      2
78   129  8  6  1  7  1  3  7  0  2      2
79   108  9  0  0  1  0  1  9  0  2      2
80   168  7  0  0  7  9  2  9  0  0      5
81    79  1  7  4  3  5  2  0  0  2      2
82    63  8  6  7  0  1  2  7  0  1      1
83    10  5  2  5  4  6  4  3  0  1      1
84    35  5  0  4  3  5  4  5  0  0      1
85    61  7  6  6  1  3  1  5  0  0      1
86    29  5  0  4  4  7  4  3  0  0      1
87   208  6  8  0  8  5  8  0  1  0      7
88   100  6  2  1  4  5  5  5  0  0      2
89   187  6  8  2  9  0  8  0  1  0      7
90   135  6  4  7  3  4  3  2  0  0      2
91   115  6  3  7  2  3  5  3  0  0      2
92   171  9  5  0  6  2  2  9  0  0      5
93    19  7  7  7  2  1  1  6  0  0      1
94     9  7  7  5  4  1  4  2  0  0      1
95    38  6  1  4  4  5  7  5  0  0      1
96   154  1  5  2  2  3  5  2  0  0      3
97    69  8  3  5  0  2  2  8  0  2      1
98    23  4  1  6  3  4  5  5  0  0      1
99   196  0  7  0  9  9  1  7  0  0      7
100   66  8  6  5  1  1  1  7  0  0      1
101   26  5  2  4  2  6  7  3  0  0      1
102   80  1  1  4  8  5  8  0  0  0      2
103  141  4  4  4  6  3  8  1  0  0      2
104   14  4  1  5  3  8  3  3  0  2      1
105  132  9  6  0  4  0  2  9  0  1      2
106  102  4  0  1  7  5  8  7  0  0      2

[107 rows x 11 columns]

### Comment on assumptions, things that do not work properly, etc.

## 1d. Divide a dataset into a training and a test set

In [47]:
# Insert the function split below
#
# Input to split:
# df: a dataframe
# testfraction: a float in the range (0,1) (default = 0.5)
#
# Output from split:
# trainingdf: a dataframe consisting of a random sample of (1-testfraction) of the rows in df
# testdf: a dataframe consisting of the rows in df that are not included in trainingdf
#
# Hint: You may use np.random.permutation(df.index) to get a permuted list of indexes where a 
#       prefix corresponds to the test instances, and the suffix to the training instances 

def ceil(n):
    res = int(n)
    return res if res == n  else res+1

def split(dff, testfraction=0.5):
    # Create two copies of dff.
    # One will be used for train df, the other for test df
    tr_df = dff.copy()
    te_df = dff.copy()

    perm = np.random.permutation(dff.index)
    frac = testfraction

    # Split the permutation array into 2 arrays s.t.
    # first one corresponds to test df indexes (tee)
    # the second to train df indexes. (trr)
    # Tot_am = total amount of rows in df, i.e. elements in perm
    # te_am = amount of rows that will be in test dataset

    tot_am = len(perm)

    te_am = ceil(frac*tot_am)
    tee = perm[:(te_am )]
    trr = perm[(te_am) :]

    # Create the new dataset dropping the rows not needed, i.e for the train
    # dataset we drop all the rows that are in perm but not in trr.
    # Same thing (but the other way around using tee) for the test dataset

    for i in perm:
        if i not in trr:
            tr_df = tr_df.drop(index=[i])
        elif i in trr:
            te_df = te_df.drop(index=[i])

    return tr_df, te_df


In [48]:
# Test your code  (leave this part unchanged)

glass_df = pd.read_csv("glass.txt")

glass_train, glass_test = split(glass_df,testfraction=0.25)

print("Training IDs:\n{}".format(glass_train["ID"].values))

print("\nTest IDs:\n{}".format(glass_test["ID"].values))

print("\nOverlap: {}".format(set(glass_train["ID"]).intersection(set(glass_test["ID"]))))


Training IDs:
[  2   3   4   5   6   8   9  10  11  12  17  18  19  22  23  26  27  28
  30  31  32  33  34  35  36  38  40  41  42  43  45  46  47  50  51  52
  54  56  57  59  60  61  62  65  67  68  70  71  72  73  74  75  76  77
  78  79  81  82  83  85  86  91  92  93  94  96  97  98 100 101 105 106
 107 108 110 111 112 113 114 115 116 117 118 119 121 122 123 124 126 127
 128 130 132 133 134 135 136 137 138 139 142 144 145 147 148 149 150 151
 154 156 157 158 159 160 161 165 166 167 168 169 170 171 172 173 175 176
 177 179 180 181 182 184 186 187 188 189 190 191 192 194 195 196 197 198
 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214]

Test IDs:
[  1   7  13  14  15  16  20  21  24  25  29  37  39  44  48  49  53  55
  58  63  64  66  69  80  84  87  88  89  90  95  99 102 103 104 109 120
 125 129 131 140 141 143 146 152 153 155 162 163 164 174 178 183 185 193]

Overlap: set()


As the use math library was not allowed, we defined a ceil function. This was needed because when definining the amount of instances in train and test dataframes it could be the case that (testfraction*(total amount of instances)) is a float, e.g 53.5. 
 Indeed, if this function was not implemented then the selection of indexes in tee and trr would have not been uniquely defined. For example, for this particular dataset if one would use int() in te_am, to get 54 rows in test dataframes one would need to define tee  as tee = perm[:te_am + 1]. So, if one sets the testfraction to 0.5 (default value) then in the two resulting dataframes one would have respectively 106 and 108 instances for test dataframe and train dataframe. This is not what one would expect. There must be 107 instances in both dataframe.

Given all this, since we can not have 53.5 instances (we need an integer), we chose to use the ceil of it. 




## 1e. Calculate accuracy of a set of predictions

In [49]:
# Insert the function accuracy below
#
# Input to accuracy:
# df: a dataframe with class labels as column names and each row corresponding to
#     a prediction with estimated probabilities for each class
# correctlabels: an array (or list) of the correct class label for each prediction
#                (the number of correct labels must equal the number of rows in df)
#
# Output from accuracy:
# accuracy: the fraction of cases for which the predicted class label coincides with the correct label
#
# Hint: In case the label receiving the highest probability is not unique, you may
#       resolve that by picking the first (as ordered by the column names) or 
#       by randomly selecting one of the labels with highest probaility.
def accuracy(df, correctlabels):
	dff = df.copy()
	if len(correctlabels)!= len(dff.index):
		 print("Error, mismatching dimensions!")
		 return None
 
	
	# Create a series where the class corresponding to the largest element
	# in a column are collected. NOTE: in case of a tie, the first class is
	# selected.
	
	maximum = dff.idxmax(axis=1)
	num = 0
	
	for i in range(len(maximum)):
		if maximum[i] == correctlabels[i]:
			num += 1
	den = len(maximum)
	
	frac = num / den

	return frac


In [50]:
# Test your code  (leave this part unchanged)

predictions = pd.DataFrame({"A":[0.5,0.5,0.5,0.25,0.25],"B":[0.5,0.25,0.25,0.5,0.25],"C":[0.0,0.25,0.25,0.25,0.5]})
predictions


A     B     C
0  0.50  0.50  0.00
1  0.50  0.25  0.25
2  0.50  0.25  0.25
3  0.25  0.50  0.25
4  0.25  0.25  0.50

In [51]:
correctlabels = ["B","A","B","B","C"]

accuracy(predictions,correctlabels) # Note that depending on how ties are resolved the accuracy may be 0.6 or 0.8

0.6

In this task, if two classes had the same largest probability for a given instance (hence over a row) we decided to label the instance by picking the first class. Indeed, for the first instance, we picked "A" as the predicted class because both classes "A" and "B" were suitible for being the predicted class.

## 2a. Create and apply one-hot encoding

In [52]:
# Insert the functions create_one_hot and apply_one_hot below
#
# Input to create_one_hot:
# df: a dataframe
#
# Output from create_one_hot:
# df: a new dataframe, where each categoric feature has been replaced by a set of binary features 
#    (as many new features as there are possible values)
# one_hot: a mapping (dictionary) from column name to a set of categories (possible values for the feature)
#
# Hint 1: First copy the input dataframe and modify the copy (the input dataframe should be kept unchanged)
# Hint 2: Consider columns of type "object" or "category" only (and which are not labeled "CLASS" or "ID")
# Hint 3: Consider creating new column names by merging the original column name and the categorical value
# Hint 4: Set all new columns to be of type "float"
# Hint 5: Do not forget to remove the original categoric feature
#
# Input to apply_one_hot:
# df: a dataframe
# one_hot: a mapping (dictionary) from column name to categories
#
# Output from apply_one_hot:
# df: a new dataframe, where each categoric feature has been replaced by a set of binary features
#
# Hint: See the above Hints



In [53]:
def create_one_hot(df):
    # Apply get_dummies to everything except "CLASS" and "ID"
       
    if "CLASS" and "ID" in df.columns:
        df_copy = df.drop(["CLASS","ID"], axis = 1) 
    elif "CLASS" in df.columns:
        df_copy = df.drop(["CLASS"], axis = 1) 
    elif "ID" in df.columns:
        df_copy = df.drop(["ID"], axis = 1) 
    else:
        df_copy = df.copy()
     
    # -------------------------------------------------------------------------
    new_df = pd.get_dummies(df_copy).astype(float) #want to have it as float    

    # Add the column with the classes and ID as the first column and last column resp.
    if "CLASS" and "ID" in df.columns:
        new1 = df["CLASS"]
        new2 = df["ID"]
        combined_df = pd.concat([new1, new_df,new2], axis=1) 
    elif "CLASS" in df.columns:
        new1 = df["CLASS"]
        combined_df = pd.concat([new1,new_df], axis=1) 
    elif "ID" in df.columns:
        new2 = df["ID"]
        combined_df =  pd.concat([new_df,new2], axis=1) 
    else:
        combined_df = df.copy()
    # -------------------------------------------------------------------------
    
    col_names = list(df_copy)
    
    # Aapping for all the columns in the original data set except the "CLASS" and "ID"
    one_hot = {}
    for i in range(len(col_names)):
        x = col_names[i]
        val = list(set(df_copy[x]))
        one_hot[x] = val

    return combined_df, one_hot



In [54]:
def apply_one_hot(df, one_hot):
    # Here we will have the test data set
    # Apply get_dummies to everything except "CLASS" and "ID"
    if "CLASS" and "ID" in df.columns:
        df_copy = df.drop(["CLASS","ID"], axis = 1) 
    elif "CLASS" in df.columns:
        df_copy = df.drop(["CLASS"], axis = 1) 
    elif "ID" in df.columns:
        df_copy = df.drop(["ID"], axis = 1) 
    else:
        df_copy = df.copy()
	
	    
    
    # Note:  the prefix will make sure we have the same columns as for the training set
    new_df = pd.get_dummies(df_copy, prefix = one_hot.keys()).astype(float) #want to have it as float  
    
    # Add the column with the classes as the first column
    if "CLASS" and "ID" in df.columns:
        new1 = df["CLASS"]
        new2 = df["ID"]
        combined_df = pd.concat([new1, new_df,new2], axis=1) 
    elif "CLASS" in df.columns:
        new1 = df["CLASS"]
        combined_df = pd.concat([new1,new_df], axis=1) 
    elif "ID" in df.columns:
        new2 = df["ID"]
        combined_df =  pd.concat([new_df,new2], axis=1) 
    else:
        combined_df = df.copy()

    
    return combined_df  


In [55]:
# Test your code  (leave this part unchanged)

tictactoe = pd.read_csv("tic-tac-toe.txt")

train_df, test_df = split(tictactoe) # Using your above function

new_train, one_hot = create_one_hot(train_df)

new_test = apply_one_hot(test_df,one_hot)
new_test

CLASS  top-left-square_b  top-left-square_o  top-left-square_x  \
0    positive                0.0                0.0                1.0   
8    positive                0.0                0.0                1.0   
14   positive                0.0                0.0                1.0   
20   positive                0.0                0.0                1.0   
21   positive                0.0                0.0                1.0   
22   positive                0.0                0.0                1.0   
23   positive                0.0                0.0                1.0   
26   positive                0.0                0.0                1.0   
28   positive                0.0                0.0                1.0   
29   positive                0.0                0.0                1.0   
31   positive                0.0                0.0                1.0   
32   positive                0.0                0.0                1.0   
34   positive                0.0                0.0                1.0   
35   positive                0.0                0.0                1.0   
37   positive                0.0                0.0                1.0   
38   positive                0.0                0.0                1.0   
40   positive                0.0                0.0                1.0   
42   positive                0.0                0.0                1.0   
44   positive                0.0                0.0                1.0   
45   positive                0.0                0.0                1.0   
47   positive                0.0                0.0                1.0   
48   positive                0.0                0.0                1.0   
51   positive                0.0                0.0                1.0   
52   positive                0.0                0.0                1.0   
54   positive                0.0                0.0                1.0   
56   positive                0.0                0.0                1.0   
59   positive                0.0                0.0                1.0   
60   positive                0.0                0.0                1.0   
61   positive                0.0                0.0                1.0   
62   positive                0.0                0.0                1.0   
..        ...                ...                ...                ...   
886  negative                1.0                0.0                0.0   
887  negative                1.0                0.0                0.0   
889  negative                1.0                0.0                0.0   
890  negative                1.0                0.0                0.0   
894  negative                1.0                0.0                0.0   
897  negative                1.0                0.0                0.0   
898  negative                1.0                0.0                0.0   
901  negative                1.0                0.0                0.0   
902  negative                1.0                0.0                0.0   
903  negative                1.0                0.0                0.0   
904  negative                1.0                0.0                0.0   
909  negative                1.0                0.0                0.0   
910  negative                1.0                0.0                0.0   
911  negative                1.0                0.0                0.0   
920  negative                1.0                0.0                0.0   
921  negative                1.0                0.0                0.0   
925  negative                1.0                0.0                0.0   
927  negative                1.0                0.0                0.0   
929  negative                1.0                0.0                0.0   
930  negative                1.0                0.0                0.0   
932  negative                1.0                0.0                0.0   
942  negative                0.0                0.0                1.0   
944  negativ

If one compares the output to the output provided, two things can be noted:
    - the rows listing is ordered: this is because the function split provides an ordered list of indexes in increasing order.
      To get a mixed order shuffle(from numpy library) could be used.
    - the number of columns is 28. This is because for top-left-square we have encoded with 3 values (for x-o-b) differently 
      from the output provided.
It is assumed that the dataset provided is categorical, hence it makes sense to apply one-hot encoding so that we get eventually numerical values. Otherwise it should be checked.

## 2b. Divide a dataset into a number of folds

In [57]:
# Insert the function folds below
#
# Input to folds:
# df: a dataframe
# nofolds: an integer greater than 1 (default = 10)
#
# Output from folds:
# folds: a list (of length = nofolds) dataframes consisting of random non-overlapping, 
#        approximately equal-sized subsets of the rows in df
#
# Hint: You may use np.random.permutation(df.index) to get a permuted list of indexes from which a 
#       prefix corresponds to the test instances, and the suffix to the training instances 

def folds(df,nofolds=10):
	
    # Create a permuted array of indexes of df
	perm = np.random.permutation(df.index)
    # Create the folds using numpy function array_split 
    # where each fold has indexes picked from per
    # The function creates nofolds folds of similar size
	folds = (np.array_split(perm, nofolds))
		
	return folds


In [58]:
# Test your code  (leave this part unchanged)

glass_df = pd.read_csv("glass.txt")

glass_folds = folds(glass_df,nofolds=5)

fold_sizes = [len(f) for f in glass_folds]

print("Fold sizes:{}\nTotal no. instances: {}".format(fold_sizes,sum(fold_sizes)))

Fold sizes:[43, 43, 43, 43, 42]
Total no. instances: 214


### Comment on assumptions, things that do not work properly, etc.

## 2c. Calculate Brier score of a set of predictions

In [59]:
# Insert the function brier_score below
#
# Input to brier_score:
# df: a dataframe with class labels as column names and each row corresponding to
#     a prediction with estimated probabilities for each class
# correctlabels: an array (or list) of the correct class label for each prediction
#                (the number of correct labels must equal the number of rows in df)
#
# Output from brier_score:
# brier_score: the average square error of the predicted probabilties 
#
# Hint: Compare each predicted vector to a vector for each correct label, which is all zeros except 
#       for at the index of the correct class. The index can be found using np.where(df.columns==l)[0] 
#       where l is the correct label.

def brier_score(df,correctlabels):
	# Create an empty list where all partial sums will be stored
	score = []
	
	# Iterate over each row according to the definition:
	# for each prediction do (pi-oi)^2 where pi is the predicted
	# probability and oi is the real one, i.e. it is 1 iff class of i is the
	# correct label o/w it is 0
	
	for i in df.index:
		# Create partial sum for each row
		summ = 0
		row = np.array(df.loc[i])
		
		# Initialize o vector for the current row. Initialized to 0, and the
		# element in position indx is set to 1, because idx is the index
		# corresponding to the correct class
		
		o = np.zeros(len(row))
		indx = np.where(df.columns==correctlabels[i])[0]
		o[indx] = 1
		for i in range(len(row)):
			summ += (row[i] - o[i])**2
		score.append(summ)
	
	# Normalize the score dividing by the number of instances = number of rows 
	# = number of prediction vectors!
	brier = sum(score)/len(df.index)
	
	return brier

In [60]:
# Test your code  (leave this part unchanged)

predictions = pd.DataFrame({"A":[0.5,0.5,0.5,0.25,0.25],"B":[0.5,0.25,0.25,0.5,0.25],"C":[0.0,0.25,0.25,0.25,0.5]})

correctlabels = ["B","A","B","B","C"]

brier_score(predictions,correctlabels)

0.5

### Comment on assumptions, things that do not work properly, etc.

## 2d. Calculate AUC of a set of predictions

In [61]:
# Insert the function auc below
#
# Input to auc:
# df: a dataframe with class labels as column names and each row corresponding to
#     a prediction with estimated probabilities for each class
# correctlabels: an array (or list) of the correct class label for each prediction
#                (the number of correct labels must equal the number of rows in df)
#
# Output from auc:
# auc: the weighted area under ROC curve
#
# Hint 1: Calculate the binary AUC first for each class label c, i.e., treating the
#         predicted probability of this class for each instance as a score; the true positives
#         are the ones belonging to class c and the false positives the rest
# Hint 2: When calculating the binary AUC, first find the scores of the true positives and then
#         the scores of the true negatives
# Hint 3: You may use a dictionary with a mapping from each score to an array of two numbers; 
#         the number of true positives with this score and the number of true negatives with this score
# Hint 4: Created a (reversely) sorted (on the scores) list of pairs from the dictionary and
#         iterate over this to additively calculate the AUC
# Hint 5: For each pair in the above list, there are three cases to consider; the no. of false positives
#         is zero, the no. of true positives is zero, and both are non-zero
# Hint 6: Calculate the weighted AUC by summing the individual AUCs weighted by the relative
#         frequency of each class (as estimated from the correct labels)

def auc(df,correctlabels):
	#Create a vector where to store partial AUCs for each class (i.e. column) 
	AUCT = [] 
	for col in df.columns:
		 # For each class create a dictionary to store {score: tp, fp}
		d = {}
		values = df[col].values
		# Create array of scores values in decreasing order
		val = sorted(values, reverse = True)  
		# For each element in val iterate and set the threshold equal to the 
		# current value and label the instances accordingly to this 
		for elem in val:
			# Current score = threshold = each value in the column iteratively
			s = elem 
			tp = 0
			fp = 0
			# For each score in the unordered list of scores check if the label is correct
			# If so, increase by 1 tp; o/w increase by 1 fp
			for j in range(len(values)):
				if values[j] == s:
					label = col
					if label == correctlabels[j]:
						tp += 1
					else:
						fp += 1
			# Append the values of fp, tp to the corresponding score
			d[s] = [tp,fp]
			
		# Implementation of AUC area algorithm for each class
		Tot_tp = 0
		Tot_fp = 0
		for elem in d:
			Tot_tp += d[elem][0]
			Tot_fp += d[elem][1]
		AUC = 0
		Cov_tp = 0
		for elem in d:
			fp = d[elem][1]
			tp = d[elem][0]
			if fp == 0:
				Cov_tp += tp
			elif tp == 0:
				AUC += (Cov_tp/Tot_tp)*(fp/Tot_fp)
			else:
				AUC += (Cov_tp/Tot_tp)*(fp/Tot_fp)+(tp/Tot_tp)*(fp/Tot_fp)/2
				Cov_tp += tp
		AUCT.append(AUC)
	
	
	#Compute the frequency vector for classes' occurence w.r.t. correct labels
	fr = []
	for col in df.columns:
		freq = correctlabels.count(col)
		freq = freq /len(correctlabels)
		fr.append(freq)
	
	# Return weighted auc = return scalar product AUCT & frequency vector
	return np.dot(fr,AUCT)

In [62]:
# Test your code  (leave this part unchanged)

predictions = pd.DataFrame({"A":[0.9,0.9,0.6,0.55],"B":[0.1,0.1,0.4,0.45]})

correctlabels = ["A","B","B","A"]

auc(predictions,correctlabels)



0.375

In [63]:
predictions = pd.DataFrame({"A":[0.5,0.5,0.5,0.25,0.25],"B":[0.5,0.25,0.25,0.5,0.25],"C":[0.0,0.25,0.25,0.25,0.5]})

correctlabels = ["B","A","B","B","C"]

auc(predictions,correctlabels)

0.8499999999999999

### Comment on assumptions, things that do not work properly, etc.